pkg/cloud-provider/cloudapi/aws/aws_security.go

```diff
@@ -590,6 +593,82 @@ func (ec2Cfg *ec2ServiceConfig) getNepheControllerManagedSecurityGroupsCloudView
 	return enforcedSecurityCloudView
 }
 
+// computeIPPermissionDelta compute the difference between new ip permissions and existing ip permissions.
+func computeIPPermissionDelta(newIpPermissions, currentIpPermissions []*ec2.IpPermission) ([]*ec2.IpPermission, []*ec2.IpPermission) {
+	ipPermissionsAdded := make([]*ec2.IpPermission, 0)
+	ipPermissionsDeleted := make([]*ec2.IpPermission, 0)
+
+	for newIdx, newIpPermission := range newIpPermissions {
+		for currentIdx, currentIpPermission := range currentIpPermissions {
+			if compareIPPermissions(newIpPermission, currentIpPermission) {
+				newIpPermissions[newIdx] = nil
+				currentIpPermissions[currentIdx] = nil
+				break
+			}
+		}
+	}
+
+	for _, IpPermission := range newIpPermissions {
+		if IpPermission != nil {
+			ipPermissionsAdded = append(ipPermissionsAdded, IpPermission)
+		}
+	}
+
+	for _, IpPermission := range currentIpPermissions {
+		if IpPermission != nil {
+			ipPermissionsDeleted = append(ipPermissionsDeleted, IpPermission)
+		}
+	}
+	return ipPermissionsAdded, ipPermissionsDeleted
+}
+
+// compareIPPermissions check if two ip permissions are equal.
+// it checks FromPort, ToPort, IpRanges, Ipv6Ranges, PrefixListIds, IpProtocol, UserIdGroupPair.GroupId.
+func compareIPPermissions(p1, p2 *ec2.IpPermission) bool {
+	if p1 == nil || p2 == nil {
+		return p1 == p2
+	}
+	if !reflect.DeepEqual(p1.FromPort, p2.FromPort) {
+		return false
+	}
+	if !reflect.DeepEqual(p1.ToPort, p2.ToPort) {
+		return false
+	}
+	if !reflect.DeepEqual(p1.IpRanges, p2.IpRanges) {
+		return false
+	}
+	if !reflect.DeepEqual(p1.Ipv6Ranges, p2.Ipv6Ranges) {
+		return false
+	}
+	if !reflect.DeepEqual(p1.PrefixListIds, p2.PrefixListIds) {
+		return false
+	}
+
+	// aws return protocol name ("tcp"), nephe uses protocol number string ("6"). A conversion to number string is needed.
+	p1Protocol, p2Protocol := p1.IpProtocol, p2.IpProtocol
+	if protocol, ok := securitygroup.ProtocolNameNumMap[*p1Protocol]; ok {
+		p1Protocol = convertToIPPermissionProtocol(&protocol)
+	}
+	if protocol, ok := securitygroup.ProtocolNameNumMap[*p2Protocol]; ok {
+		p2Protocol = convertToIPPermissionProtocol(&protocol)
+	}
+	if *p1Protocol != *p2Protocol {
+		return false
+	}
+
+	groupIds := make(map[string]bool)
+	for _, id := range p1.UserIdGroupPairs {
+		groupIds[*id.GroupId] = true
```

Comment:
I believe this is addressgroup AG ID. Can there be case we have more than 1 UserIdGroup Pair?